### ISSUE TITLE
Unable to move items under certain scenarios

### ISSUE THREAD

**Issue by CLGitDev** [angular-ui/ui-sortable#387]
I'm working on an AngularJS project with a feature that users can move items across arrays. Only one array is shown on initialization; when users drag an item and mouseover on the desired array name, that particular array will be shown instead and users can drop the item there.

Below is a codepen project to illustrate the feature: http://codepen.io/anon/pen/PPWpQV

I have gotten it kinda working, however there are two scenarios which I'm experiencing problems:

When there are three arrays, I'm able to move items across them (even to the empty array). However the last item in an array cannot be moved. Besides, the droppable area seems to be dependent on the source array. For instance, if the source array has 2 items while the target array has 5, the placeholder will only appear when I mouseover the cursor on the first 2 elements of the target array.

When there are two arrays only and one is empty, I'm unable to move any items at all. (Try commenting out one of the non-empty array.)

If all arrays are shown on initialization instead of show on mouseover (i.e. removing the ng-show attribute), the sortable feature work seamlessly.

I've spent some time trying to figure it out but didn't managed to get my head around. Any advice are much appreciated!

**Comment by thgreasi** [comment]
Hmmm.
I've seen some reports related to the initial visibility state of the
sortables. This should be on the jquery-ui part and not in our wrapper
directive. Can you try doing a  .sortable('refresh') after changing the
visibility of each directive?

https://api.jqueryui.com/sortable/#method-refresh

On Wed, Sep 30, 2015, 16:44 CLGitDev <notifications@github.com> wrote:

> I'm working on an AngularJS project with a feature that users can move
> items across arrays. Only one array is shown on initialization; when users
> drag an item and mouseover on the desired array name, that particular array
> will be shown instead and users can drop the item there.
>
> Below is a codepen project to illustrate the feature:
> http://codepen.io/anon/pen/PPWpQV
>
> I have gotten it kinda working, however there are two scenarios which I'm
> experiencing problems:
>
> When there are three arrays, I'm able to move items across them (even to
> the empty array). However the last item in an array cannot be moved.
> Besides, the droppable area seems to be dependent on the source array. For
> instance, if the source array has 2 items while the target array has 5, the
> placeholder will only appear when I mouseover the cursor on the first 2
> elements of the target array.
>
> When there are two arrays only and one is empty, I'm unable to move any
> items at all. (Try commenting out one of the non-empty array.)
>
> If all arrays are shown on initialization instead of show on mouseover
> (i.e. removing the ng-show attribute), the sortable feature work seamlessly.
>
> I've spent some time trying to figure it out but didn't managed to get my
> head around. Any advice are much appreciated!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular-ui/ui-sortable/issues/387>.
>
-- 
*Thodoris Greasidis*
Computer, Networking &
Communications Engineer

**Comment by thgreasi** [comment]
By the way, that's an issue report with great description and level of
details.

On Wed, Sep 30, 2015, 18:33 Thodoris Greasidis <thgreasi@gmail.com> wrote:

> Hmmm.
> I've seen some reports related to the initial visibility state of the
> sortables. This should be on the jquery-ui part and not in our wrapper
> directive. Can you try doing a  .sortable('refresh') after changing the
> visibility of each directive?
>
> https://api.jqueryui.com/sortable/#method-refresh
>
> On Wed, Sep 30, 2015, 16:44 CLGitDev <notifications@github.com> wrote:
>
>> I'm working on an AngularJS project with a feature that users can move
>> items across arrays. Only one array is shown on initialization; when users
>> drag an item and mouseover on the desired array name, that particular array
>> will be shown instead and users can drop the item there.
>>
>> Below is a codepen project to illustrate the feature:
>> http://codepen.io/anon/pen/PPWpQV
>>
>> I have gotten it kinda working, however there are two scenarios which I'm
>> experiencing problems:
>>
>> When there are three arrays, I'm able to move items across them (even to
>> the empty array). However the last item in an array cannot be moved.
>> Besides, the droppable area seems to be dependent on the source array. For
>> instance, if the source array has 2 items while the target array has 5, the
>> placeholder will only appear when I mouseover the cursor on the first 2
>> elements of the target array.
>>
>> When there are two arrays only and one is empty, I'm unable to move any
>> items at all. (Try commenting out one of the non-empty array.)
>>
>> If all arrays are shown on initialization instead of show on mouseover
>> (i.e. removing the ng-show attribute), the sortable feature work seamlessly.
>>
>> I've spent some time trying to figure it out but didn't managed to get my
>> head around. Any advice are much appreciated!
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/angular-ui/ui-sortable/issues/387>.
>>
> --
> *Thodoris Greasidis*
> Computer, Networking &
> Communications Engineer
>
-- 
*Thodoris Greasidis*
Computer, Networking &
Communications Engineer

**Comment by CLGitDev** [comment]
@thgreasi I thought my passage is too lengthy to read. Thanks a lot for your advice, that's a life saver!

I've tried the refresh method but it didn't work at first. Then I read your advice again and realized that I've missed out a very important point you mentioned - the method has to be called <b>after changing the visibility of each directive</b>!

What I've done wrong was - I called the refresh method straightaway after changing the visible_array_id (which is bound to the ng-show directive). The digest cycle hasn't been completed by that time, so the ng-show directive hasn't been evaluated nor the DOM element hasn't been updated. Realizing this, I wrapped the refresh method with a timeout function to defer it until DOM update finishes. It is now working seamlessly.

I've modified my codepen project to incorporate the logic above:
http://codepen.io/anon/pen/MaJqZe
